test/metatx/ERC2771Forwarder.t.sol

```diff
@@ -162,4 +227,79 @@ contract ERC2771ForwarderTest is Test {
         assertEq(address(_erc2771Forwarder).balance, initialBalance);
         assertEq(refundReceiver.balance, expectedRefund);
     }
+
+    function testVerifyTamperedValues(uint8 tamper) public {
+        tamper = uint8(bound(tamper, 0, 4));
+
+        assertFalse(
+            _erc2771Forwarder.verify(
+                _tamperedValues(
+                    tamper,
+                    _forgeRequestData({
+                        value: 0,
+                        nonce: 0,
+                        deadline: uint48(block.timestamp + 1),
+                        data: abi.encodeCall(CallReceiverMock.mockFunction, ())
+                    })
+                )
+            )
+        );
+    }
+
+    function testExecuteTamperedValues(uint8 tamper) public {
+        tamper = uint8(bound(tamper, 0, 4));
+
+        ERC2771Forwarder.ForwardRequestData memory request = _tamperedExecute(tamper, 0);
+
+        _erc2771Forwarder.execute{value: request.value}(request);
+    }
+
+    function testExecuteBatchTamperedValuesZeroReceiver(uint8 tamper) public {
+        tamper = uint8(bound(tamper, 0, 4));
+        uint256 batchSize = 3;
+
+        ERC2771Forwarder.ForwardRequestData[] memory batchRequestDatas = new ERC2771Forwarder.ForwardRequestData[](
+            batchSize
+        );
+        for (uint256 i = 0; i < batchSize; ++i) {
+            batchRequestDatas[i] = _forgeRequestData({
+                value: 0,
+                nonce: i,
+                deadline: uint48(block.timestamp + 1),
+                data: abi.encodeCall(CallReceiverMock.mockFunction, ())
+            });
+        }
+
+        batchRequestDatas[1] = _tamperedExecute(tamper, 1);
+
+        _erc2771Forwarder.executeBatch{value: batchRequestDatas[1].value}(batchRequestDatas, payable(address(0)));
+    }
+
+    function testExecuteBatchTamperedValues(uint8 tamper) public {
+        tamper = uint8(bound(tamper, 0, 4));
+        uint256 batchSize = 3;
+
+        ERC2771Forwarder.ForwardRequestData[] memory batchRequestDatas = new ERC2771Forwarder.ForwardRequestData[](
+            batchSize
+        );
+        for (uint256 i = 0; i < batchSize; ++i) {
+            batchRequestDatas[i] = _forgeRequestData({
+                value: 0,
+                nonce: i,
+                deadline: uint48(block.timestamp + 1),
+                data: abi.encodeCall(CallReceiverMock.mockFunction, ())
+            });
+        }
+
+        batchRequestDatas[1] = _tamperedValues(tamper, batchRequestDatas[1]);
+
+        vm.deal(address(this), batchRequestDatas[1].value);
+
+        vm.recordLogs();
+        _erc2771Forwarder.executeBatch{value: batchRequestDatas[1].value}(batchRequestDatas, payable(address(0xebe)));
+
+        Vm.Log[] memory entries = vm.getRecordedLogs();
+        assertEq(entries.length, 2);
+        assertEq(entries[1].topics[0], keccak256("ExecutedForwardRequest(address,uint256,bool)"));
```

Comment:
Although this works, I feel there should be a cleaner way

Reply:
i'm not aware of a cleaner because `vm.expectEmit` doesn't allow to catch not-expected events (i.e. it's able to catch the lack of events but not the opposite afaik)

Reply:
fwiw this approach mimicks the behavior from the [hardhat tests](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/5258/files#diff-29fad6b26849463f244fab4e002bfb9306c154b53ef8995d5a5d546de6eacc0fL352)

Reply:
The hardhat test you link looks like an old thing from the web3js times. If we were to (re)write this tests today we would use the `expect().to.not.emit()` syntax from the new helpers.

Reply:
gotcha, replaced with a more concrete check https://github.com/OpenZeppelin/openzeppelin-contracts/pull/5258/commits/598a09ba2f31416e23af1bb5be8db486cd526d54